File: app/src/main/java/com/solutis/filmes/retrofit/callback/BaseCallback.java
package com.solutis.filmes.retrofit.callback;

import retrofit2.Call;
import retrofit2.Response;
import retrofit2.internal.EverythingIsNonNull;

public class BaseCallback<T> implements retrofit2.Callback<T> {

    private final CallBackResponse<T> callback;

    public BaseCallback(CallBackResponse<T> callback) {
        this.callback = callback;
    }

    @Override
    @EverythingIsNonNull
    public void onResponse(Call<T> call, Response<T> response) {
        if(!response.isSuccessful()){
            this.callback.fail(response.message());
            return;
        }


        this.callback.success(response.body());
    }

    @Override
    @EverythingIsNonNull
    public void onFailure(Call<T> call, Throwable t) {
        this.callback.fail(t.getMessage());
    }

    public interface CallBackResponse<T> {
        void success(T result);
        void fail(String error);
    }
}

File: app/src/main/java/digital/solutis/viciadosemfilmes/retrofit/callback/BaseCallback.kt
package digital.solutis.viciadosemfilmes.retrofit.callback

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.internal.EverythingIsNonNull


class BaseCallback<T>(private val callback: ResponseCallback<T>) :
    Callback<T> {
    @EverythingIsNonNull
    override fun onResponse(
        call: Call<T>,
        response: Response<T>
    ) {
        if (response.isSuccessful) {
            val result = response.body()
            if (result != null) {
                callback.whenSuccessful(result)
            }
        } else {
            callback.whenFails("Resposta não sucedida.")
        }
    }

    @EverythingIsNonNull
    override fun onFailure(call: Call<T>, t: Throwable) {
        callback.whenFails("Falha de comunicação: " + t.message)
    }

    interface ResponseCallback<T> {
        fun whenSuccessful(result: T)
        fun whenFails(error: String?)
    }

}
